src/commands/mod.rs

```diff
@@ -2346,6 +2424,143 @@ Please use `jj new 'all:x|y'` instead of `jj new --allow-large-revsets x y`.",
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let edit = args.edit;
+    let amount = args.amount;
+    assert!(amount >= 1);
+    let current_wc_id = workspace_command
+        .get_wc_commit_id()
+        .ok_or_else(|| user_error("This command requires a working copy"))?;
+    let current_wc = workspace_command.repo().store().get_commit(current_wc_id)?;
+    let current_short = short_commit_hash(current_wc.id());
+    // If we're editing, start at the working-copy commit.
+    // Everything else starts from our direct parent.
+    let start_id = if edit {
+        current_wc_id
+    } else {
+        match current_wc.parent_ids() {
+            [parent_id] => parent_id,
+            _ => return Err(user_error("Cannot run `jj next` on a merge commit")),
+        }
+    };
+    let target_expression = RevsetExpression::commit(start_id.clone())
+        .descendants_at(amount.try_into().unwrap())
+        // Negate the current working copy.
+        .minus(&RevsetExpression::commit(current_wc_id.clone()));
+    let targets: Vec<Commit> = target_expression
+        .resolve(workspace_command.repo().as_ref())?
+        .evaluate(workspace_command.repo().as_ref())?
+        .iter()
+        .commits(workspace_command.repo().store())
+        .take(2)
+        .try_collect()?;
+    let target = match targets.as_slice() {
+        [target] => target,
+        [] => {
+            // TODO: How to handle the next onto new commit part?
+            //
+            // See the comment below.
+            // If the parent is the last commit in the repository and we're not editing,
+            // just move the working-copy commit to the end. This is just a wrapped `jj
+            // new`.
+            return Err(user_error("No target_commit"));
+        }
+        _ => {
+            // TODO(#NNN) We currently cannot deal with multiple children, which result
+            // from branches. Prompt the user for resolution.
+            return Err(user_error("Ambiguous target commit"));
+        }
+    };
+    workspace_command.check_rewritable(target)?;
+    let target_short = short_commit_hash(target.id());
+    let mut tx = workspace_command.start_transaction("");
+    // We're editing, just move to the target commit.
+    if edit {
+        tx.set_description(&format!("next: {current_short} -> editing {target_short}"));
+        tx.edit(target).unwrap();
+        tx.finish(ui)?;
+        return Ok(());
+    }
+    tx.set_description(&format!("next: {current_short} -> {target_short}"));
+    let merged_tree = merge_commit_trees(tx.repo(), &[target.clone()])?;
+    // Move the working-copy commit to the new parent.
+    let new_wc_revision = tx
+        .mut_repo()
+        .new_commit(
+            command.settings(),
+            vec![target.id().clone()],
+            merged_tree.id().clone(),
+        )
+        .write()?;
+    tx.edit(&new_wc_revision).unwrap();
```

Comment:
Perhaps, `tx.check_out()` can be used here.

`cmd_checkout()` does that manually, but that's probably because `jj checkout` can also set commit description.

Reply:
I missed that during the refactor, martin also suggested it.

Reply:
done.